### PR TITLE
Improve `Database.query*` methods

### PR DESCRIPTION
**Pull request by alalazo**
This PR improves a few aspects of the `Database.query*` methods. Follows from #47102

**Modifications**
- [x] Added type hints to all `query*` methods
- [x] Inlined docstrings, as they behave better with tooling this way (e.g. quick help can be displayed) 
- [x] Changed weird default `any` to `None` (as it can be type-hinted)
- [x] `Database._query` now pre-filters on hashes to be considered for iteration
- [x] Fixes a bug where the `--origin` option of uninstall had no effect
- [x] Fixes a bug in `Database._query` that doesn't apply other selection mechanism if a spec

### CODE REVIEW

**Comment by haampie** [comment]
Looks better. The only downside of predicate_fn that occurred to me is that it makes the query opaque which would have to be written explicitly in SQL would we ever use an actual database. In that sense `db.specific_query(...)` would be better.

Also I wonder if it makes sense to move `query_local(...)` to `query(local=True)` and use that arg in other places to reduce the number of methods.

**Comment by alalazo** [comment]
> Also I wonder if it makes sense to move query_local(...) to query(local=True) and use that arg in other places to reduce the number of methods.

Looking closely at the method signature we have `install_tree="local"` that already works that way. That argument imo deserves to be an enum. It's a bit convoluted that it can accept also paths, but we can deal with it I guess.

**Comment by alalazo** [comment]
> Also I wonder if it makes sense to move query_local(...) to query(local=True) and use that arg in other places to reduce the number of methods.

I agree with this, but given that there are 54 calls over 17 files (according to IDE), would you be fine attempting it in a following PR?

**Comment by haampie** [comment]
Since `predicate_fn` is likely to be removed as soon as we have a different storage backend, I think it's better to stick to `hashes`, `origin`, etc kwargs. It's a bit arbitrary anyways that tons of "predicates" are already explicitly part of the loop body except those.

**Comment by alalazo** [comment]
> Since predicate_fn is likely to be removed as soon as we have a different storage backend, 

I don't think it needs to be removed, but I agree that `hashes=`, and perhaps `origin=`, might be optimized if we keep it as an argument.

File: lib/spack/spack/cmd/uninstall.py [hunk 1]
```diff
@@ -213,7 +215,7 @@ def get_uninstall_list(args, specs: List[spack.spec.Spec], env: Optional[ev.Envi
 
     # Gets the list of installed specs that match the ones given via cli
     # args.all takes care of the case where '-a' is given in the cli
-    matching_specs = find_matching_specs(env, specs, args.all)
+    matching_specs = find_matching_specs(env, specs, args.all, origin=args.origin)
```
Comment (alalazo) [inline lib/spack/spack/cmd/uninstall.py:217]: For the record, this one is a bug on develop

**Comment by haampie** [comment]
I think `collections.abc.Collection` is the right type for `hashes`.

Actually it should be improved performance wise, and then Collection is not necessary, plus a bug can be fixed in the meantime:

Iterate either over `_data.values()`, or `(_data[h] for h in hashes if h in _data)` or `[self._data[h]] if h in self._data else []` where `h` is a concrete input spec. Then filters apply for the latter too.

**Review by haampie — APPROVED** [review]

**Pull request merged**